tests/e2e/tests/runtime-upgrade.test.ts

```diff
@@ -73,12 +72,14 @@ describe('runtime upgrade', async () => {
     await api.tx.sudo
       .sudoUncheckedWeight(
         api.tx.system.setCode('0x' + code.toString('hex')),
-        0
+        {}
       )
       .signAndSend(alice);
 
     // Do block production.
     await dev.newBlock({ count: 2 });
+    // wait a bit for pjs/api to update
+    await new Promise((r) => setTimeout(r, 1000));
```

Comment:
How come this is needed if you await for the block production above?
Doesn't it the `newBlock` function returns prematurely?

Reply:
Alice is sending both tx and some how it needed a bit of time to sync nonce

Reply:
Just to check if I understand correctly - `remark` is executed, followed by `setCode`, two new blocks are produced, then we wait for the internal chopsticks state to _stabilize/sync_, before proceeding forward?

I don't know how chopsticks works internally, but should it return from the `newBlock` if state isn't synced yet? Maybe it's a dumb question 🤷

Reply:
once block is build  it will resolve and announce storage changes async for listeners (like ApiPromise). But because we're immediately creating a new tx (aslo because everything is running on a single js thread event loop) [nextNounce](https://github.com/polkadot-js/api/blob/5a4d11cb39572ee66b55d4baccf7872d53ab0bba/packages/api-derive/src/tx/signingInfo.ts#L25C10-L25C21) may not be reflected immediately so second tx was build with old nonce resulting in invalid tx.

Reply:
even a 200ms will be enough but I added 1sec to be sure

Reply:
This appears to happen only when we perform an upgrade, this is making me think why 🤔, need to debug it

Reply:
@Dinonard Initially I thought it was sending only nonce but it's not nonce, it's spec_version. When runtime upgrades happen, it takes a few millis for ApiPromise to receive metadata from chopsticks (because it extracts it from runtime itself, runtime call needs to happen) but it doesn't wait but instead uses what it has (old spec_version) and making an invalid tx